editor/shader_create_dialog.cpp

```diff
@@ -0,0 +1,665 @@
+}
+
+String ShaderCreateDialog::_validate_path(const String &p_path, bool p_file_must_exist) {
+	String p = p_path.strip_edges();
+
+	if (p == "") {
+		return TTR("Path is empty.");
+	}
+	if (p.get_file().get_basename() == "") {
+		return TTR("Filename is empty.");
+	}
+
+	p = ProjectSettings::get_singleton()->localize_path(p);
+	if (!p.begins_with("res://")) {
+		return TTR("Path is not local.");
+	}
+
+	DirAccess *d = DirAccess::create(DirAccess::ACCESS_RESOURCES);
+	if (d->change_dir(p.get_base_dir()) != OK) {
+		memdelete(d);
+		return TTR("Invalid base path.");
+	}
+	memdelete(d);
+
+	/* Does file already exist */
+	DirAccess *f = DirAccess::create(DirAccess::ACCESS_RESOURCES);
+	if (f->dir_exists(p)) {
+		memdelete(f);
+		return TTR("A directory with the same name exists.");
+	} else if (p_file_must_exist && !f->file_exists(p)) {
+		memdelete(f);
+		return TTR("File does not exist.");
+	}
+	memdelete(f);
+
+	/* Check file extension */
+	String extension = p.get_extension();
+	List<String> extensions;
+
+	// get all possible extensions for shader
+	for (int l = 0; l < SHADER_TYPE_MAX; l++) {
+		for (List<String>::Element *E = language_data[l].extensions.front(); E; E = E->next()) {
+			extensions.push_back(E->get());
+		}
+	}
+
+	ShaderTypeData data = language_data[language_menu->get_selected()];
+
+	bool found = false;
+	bool match = false;
+	int index = 0;
+	for (List<String>::Element *E = extensions.front(); E; E = E->next()) {
+		if (E->get().nocasecmp_to(extension) == 0) {
+			//FIXME (?) - changing language this way doesn't update controls, needs rework
+			//language_menu->select(index); // change Language option by extension
+			found = true;
+			if (E->get() == data.default_extension) {
+				match = true;
+			}
+			break;
+		}
+		index++;
+	}
+
+	if (!found) {
+		return TTR("Invalid extension.");
+	}
+	if (!match) {
+		return TTR("Wrong extension chosen.");
+	}
+
+	/* Let ScriptLanguage do custom validation */
+	String path_error = ScriptServer::get_language(language_menu->get_selected())->validate_path(p);
+	if (path_error != "") {
+		return path_error;
+	}
+
+	/* All checks passed */
+	return "";
+}
+
+void ShaderCreateDialog::_msg_script_valid(bool valid, const String &p_msg) {
+	error_label->set_text("- " + p_msg);
+	if (valid) {
+		error_label->add_theme_color_override("font_color", gc->get_theme_color("success_color", "Editor"));
+	} else {
+		error_label->add_theme_color_override("font_color", gc->get_theme_color("error_color", "Editor"));
+	}
+}
+
+void ShaderCreateDialog::_msg_path_valid(bool valid, const String &p_msg) {
+	path_error_label->set_text("- " + p_msg);
+	if (valid) {
+		path_error_label->add_theme_color_override("font_color", gc->get_theme_color("success_color", "Editor"));
+	} else {
+		path_error_label->add_theme_color_override("font_color", gc->get_theme_color("error_color", "Editor"));
+	}
+}
+
+void ShaderCreateDialog::_update_dialog() {
+	/* "Add Script Dialog" GUI logic and script checks. */
```

Comment:
I think the comments in this method could be removed, they just repeat what you can already tell from the code (not to mention that the Script Dialog copy-paste looks bad xd)

Reply:
Done